gas.md

```diff
@@ -0,0 +1,554 @@
+* `ADD RREG wREG1 wREG2`
+  - Computation requires pairwise addition of the registers' words so:
+    ```hs
+    computationCost(ADD rREG wREG1 wREG2) = constBaseAdd * wordCost * m
+    ```
+  - Size of result is about the maximum of the two
+    ```hs
+    estimatedResultSize(ADD RREG wREG1 wREG2) = m + 1
+      where m = max (registerSize wREG1) (registerSize wREG2)
+    ```
+* `SUB rREG wREG1 wREG2`
+  - Subtraction has the same complexity as addition
+    ```hs
+    computationCost(SUB rREG wREG1 wREG2) = constBaseSub * wordCost * m
+    ```
+  - Size of result is about the maximum of the two
+    ```hs
+    estimatedResultSize(SUB rREG wREG1 wREG2) = m + 1
+      where m = max (registerSize wREG1) (registerSize wREG2)
+    ```
+* `MUL rREG wREG1 wREG2`
+  -  Multiplication complexity depends on the algorithm used.
+    ```hs
+    computationCost(MUL rREG wREG1 wREG2)
+      | isZero wREG1 || isZero wREG2    = 2 * bitCost + wordCost
+      | otherWise                       = mulCost l1 l2
+    ```
+  - Size of result is the sum of sizes of the operands, except when one of them is 0
+    ```hs
+    estimatedResultSize(MUL rREG wREG1 wREG2)
+      | isZero wREG1 || isZero wREG2 = 0
+      | otherwise = l1 + l2
+      where l1 = registerSize wREG1
+            l2 = registerSize wREG2
+    ```
+* `DIV rREG wREG1 wREG2`
+  - if the value of `wREG2` is 0, or the second is larger than the first,
+    then result is 0; otherwise, we approximate the cost to that of the
+    basecase division method.
+    ```hs
+    computationCost(MUL rREG wREG1 wREG2)
+      | isZero wREG2 = bitCost + wordCopyCost
+      | l2 > l1      = lenCost + wordCopyCost
+      | otherwise    = divCost l1 l1
+    ```
+  - if the result is 0 its size is 1, otherwise it's about the difference
+    between sizes
+    ```hs
+    estimatedResultSize(MUL rREG wREG1 wREG2)
+      | isZero wREG2 = 1
+      | l2 > l1      = 1
+      | otherwise    = l1 - l2 + 1
+      where l1 = registerSize wREG1
+            l2 = registerSize wREG2
+    ```
+* `MOD rREG wREG1 wREG2`
+  - The complexity of `MOD` is similar to that of `DIV`
+    ```hs
+    computationCost(MUL rREG wREG1 wREG2)
+      | isZero wREG2 = bitCost + wordCopyCost
+      | l2 > l1      = lenCost + l1*wordCopyCost
+      | otherwise    = divCost l1 l2
+    ```
+  - if the result is 0 its size is 1, otherwise it's the minimum of the
+    operands sizes
+    ```hs
+    estimatedResultSize(MUL rREG wREG1 wREG2)
+      | isZero wREG2 = 1
+      | otherwise    = min l1 l2
+      where l1 = registerSize wREG1
+            l2 = registerSize wREG2
+    ```
+* `EXP rREG wREG1 wREG2`
+  - Exponentiation is done by repeatedly squaring `WREG1` and multiplying those
+    intermediate results for the non-zero bits of `WREG2`.
+    ```hs
+    computationCost(EXP rREG wREG1 wREG2)
+      | isZero wREG2 = bitCost + wordCopyCost
+      | otherwise    = constExp * wordCost * (0.5 * approxKara l * (approxKara e - 1) + 2 * l * (e - 1)))
```

Comment:
For exponentiation a reasonable estimate needs to use the value of `e`. For non-modular exponentiation we could maybe just immediately report out of gas if `e` is above some maximum length and not worry about the variable size.

For `EXPMOD` much larger exponents may be tractable. We may just have to do a two-phase gas check, first testing there is enough gas to evaluate a formula involving `e` and then doing the full cost calculation.

Reply:
I'm not sure I understand what happens here, so I'll wait a bit to see if Traian will answer. From the general shape of the formula, I guess that one should have ((registersize e)-1) instead of (e-1), but can't tell for sure right now. The same below.

Reply:
I agree with @bmmoore that we could do a two-phase check, and also include the price for the checks in the overall gas costs.

Reply:
I would take a look at what EVM did. They had a way of computing the gas cost of the exponent based on measuring the bit width of the exponent and doing bitwise arithmetic on the lower-order 32 bytes of the exponent. I think that if we can do something like that, it will be much better than trying to do arbitrary precision arithmetic, even though it may make the gas cost higher.